### PR TITLE
Add registration mechanism for aoti model runner

### PR DESCRIPTION
**Pull request by EikanWang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126883
* __->__ #131638

Current AOTI model runner has supported CUDA and CPU. However, in terms of a particular out-of-tree backend, it is not easier to support the feature.

This PR intends to provide a registration mechanism to support this case by providing two: `RegisterAOTIModelRunner` and `getAOTIModelRunnerRegistry`.

- `RegisterAOTIModelRunner` is used to register a function(`AOTIModelRunnerABC`) to create a `AOTIModelContainerRunner`. The function signature is as follows.

    ```C++
    using AOTIModelRunnerABC = std::shared_ptr<AOTIModelContainerRunner> (*)(
        const std::string& model_so_path,
        size_t num_models,
        const std::string& device_str,
        const std::string& bin_dir);
    ```
- `getAOTIModelRunnerRegistry` is used to get all the registered backends.

In terms of a new backend, it needs to define its `AOTIModelContainerRunner` class and then register a `AOTIModelRunnerABC` function to `aoti` to create its `AOTIModelContainerRunner`.

### CODE REVIEW

**Comment by EikanWang** [comment]
Will add test cases later.

File: torch/csrc/inductor/aoti_runner/model_container_runner.h [hunk 1]
```diff
@@ -82,5 +82,27 @@ class TORCH_API AOTIModelContainerRunner {
   std::unique_ptr<torch::aot_inductor::ProxyExecutor> proxy_executor_;
 };
 
+using CreateAOTIModelRunnerFunc = std::shared_ptr<AOTIModelContainerRunner> (*)(
+    const std::string& model_so_path,
+    size_t num_models,
+    const std::string& device_str,
+    const std::string& bin_dir);
```
Comment (EikanWang) [inline torch/csrc/inductor/aoti_runner/model_container_runner.h:89]: The signature of this function pointer is aligned with the constructor of CUDA `AOTIModelContainerRunnerCUDA`.

**Review by desertfire — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**